### PR TITLE
Add support for tasks

### PR DESCRIPTION
**Pull request by nt1m**

### CODE REVIEW

**Comment by nt1m** [comment]
@nt1m: review?(@jankeromnes)

File: lib/boot.js [hunk 1]
```diff
@@ -345,3 +346,10 @@ exports.registerDockerClient = function (next) {
     next();
   });
 };
+
+exports.loadTasks = function (callback) {
```
Comment (bnjbvr) [inline lib/boot.js]: nit: common pattern is to name the argument `next` here

File: lib/tasks.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
+const db = require('./db');
+const log = require('./log');
+let taskCounter = 0; // Used to generate unique task IDs
+
+exports.tasks = db.get('tasks');
```
Comment (bnjbvr) [inline lib/tasks.js]: General comment for this file: why are so many things exported and not used outside this module? I'd rather have them unexported until they're actually used elsewhere. This is also a big confusing in my opinion to export fields of the database file.

File: lib/tasks.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+exports.taskTypes = new Map();
+
+exports.check = function () {
+  const now = Date.now();
+  for (const task of Object.values(exports.tasks)) {
+    if (now > Number(task.date)) {
```
Comment (bnjbvr) [inline lib/tasks.js]: nit: yoda condition: please invert LHS and RHS and replace `>` by `<`

File: lib/tasks.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  return task;
+};
+
+exports.remove = function (id) {
+  delete exports.tasks[id];
+  db.save();
+};
+
+exports.execute = function ({ type, data, id }) {
+  const task = exports.taskTypes.get(type);
```
Comment (bnjbvr) [inline lib/tasks.js]: It seems that `task` is actually a function, so `taskTypes` sounds poorly named. Can you think of a better name here?

File: lib/tasks.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+  log('[ok] Task', id, 'executed', data);
+  task(data);
+  exports.remove(id);
```
Comment (bnjbvr) [inline lib/tasks.js]: Can a task be fallible? If so, shouldn't it use a callback form instead, and the task be removed only if it succeeded?

Comment (bnjbvr) [inline lib/tasks.js]: Also, can a task remove itself or add new tasks to the list of pending tasks? There should probably be a comment explaining this somewhere in this file.

File: lib/tasks.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    throw new Error('[fail] task', task, 'already exists');
+  }
+
+  exports.taskTypes.set(type, task);
+};
+
+exports.add = function (date, type, data) {
+  taskCounter++;
+
+  const msSince1970 = date.getTime();
```
Comment (bnjbvr) [inline lib/tasks.js]: nit: rename `msSince1970` to `timestamp` maybe?

File: lib/tasks.js [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+  for (const task of Object.values(exports.tasks)) {
+    if (now > Number(task.date)) {
+      exports.execute(task);
+    }
+  }
+};
+
+exports.addType = function (type, task) {
+  if (exports.taskTypes.has(type)) {
+    throw new Error('[fail] task', task, 'already exists');
```
Comment (bnjbvr) [inline lib/tasks.js]: nit: pretty sure the args to the `Error` ctor aren't doing what you expect here, so you'll need to use a template string here instead.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error

**Comment by nt1m** [comment]
@bnjbvr Addressed your comments, thanks for the review!

File: lib/tasks.js (hunk 2 above)
Comment (bnjbvr) [inline lib/tasks.js]: (not addressed)

File: lib/tasks.js (hunk 6 above)
Comment (bnjbvr) [inline lib/tasks.js]: (still present in the template string next line)

File: lib/tasks.js (hunk 4 above)
Comment (bnjbvr) [inline lib/tasks.js]: (not addressed)

File: lib/tasks.js (hunk 5 above)
Comment (bnjbvr) [inline lib/tasks.js]: (not addressed)

**Comment by bnjbvr** [comment]
@nt1m Some previous comments have not been addressed, can you comment in the code or at least in a Github comment the choices you've made?

**Comment by jankeromnes** [comment]
Hello! If you don't mind, I'd like to take over this feature in #236, so I'll close this unmerged pull request. Sorry for duplicating this great work!

**Pull request closed without merging**